Fix ancient blocks sync

ascjones opened the pull request:
Fixes #9300, fixes #7008 and fixes #9306. Rel #9407.

*Edit: there were several issues here*

## TLDR;
- Ancient blocks queue full, download resets, multiple stale responses cause download rounds with 0 blocks imported, retracting to block `#0`
- After retracting to `#0`, can download a non canonical set of blocks 
- It can get stuck on the head of a non canonical chain, when it should retract to find a common block and resume the sync.

### Retraction after queue full

1. `BlockDownloader` in `State::Blocks` [requesting headers](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/block_sync.rs#L452) for subchains: 
2. Ancient blocks [queue gets full](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/block_sync.rs#L516) and the [old blocks download is restarted](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/chain/mod.rs#L641)
3. `BlockDownloader` in `State::ChainHead`. In flight header response from abandoned round arrives, and [resets subchain headers](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/block_sync.rs#L287) to consecutive headers (see the todo comment for a clue!). These headers may be orphaned from their parents which were discarded after the queue became full above.
4. Then we get multiple sync rounds in quick succession which import 0 blocks [because of `UnknownParent`](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/block_sync.rs#L509) after, causing the sync to quickly [retract to 0](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/block_sync.rs#L387). This was also being caused by stale body responses.

The fix is to expire all pending requests when resetting the sync.

### Getting stuck on a non canonical block

1. Some peers have bad blocks in the range `#1..#57`, probably propogated by the above bug. Also around the fork block `#1920000` it was getting stuck on some non canon blocks.
2. The hash gets [stuck](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/blocks.rs#L566) in the list of heads so it never gets [empty](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/blocks.rs#L396) and the round never [completes](https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/block_sync.rs).

The fix is to reset the sync if we receive consecutive header responses for a requesetd hash with no useful headers that advance the subchain.

### Further improvements

The ancient blocks queue still fills up fairly frequently, resulting in many downloaded blocks being discarded. I will address this in a future PR, by pausing the download of old blocks similar to the normal sync. For now though this no longer triggers the retraction, and the ancient blocks sync now completes.

I also have some further improvements/refactorings for this code stashed away for future PRs. But trying to keep this PR to the minimum that will actually get the ancient block sync working again.

andresilva commented on ethcore/sync/src/block_sync.rs:
```diff
@@ -515,6 +528,7 @@ impl BlockDownloader {
 				},
 				Err(BlockImportError(BlockImportErrorKind::Queue(QueueErrorKind::Full(limit)), _)) => {
 					debug!(target: "sync", "Block import queue full ({}), restarting sync", limit);
+					bad = true;
```
This is done to trigger resetting the sync right? Do you think we should return a different error type other than `BlockDownloaderImportError::Invalid`? (And also handle it in `ChainSync::collect_blocks`).

ascjones replied:
Yes I agree and I had previously done that. However my idea with this initial PR was to do the minimum amount to fix this bug.

However looks like this fix doesn't work entirely so I will add that back while waiting for another mainnet sync...

ascjones commented:
Unfortunately this does not appear to fix the issue when syncing mainnet. I'm currently investigating.

ascjones commented on ethcore/sync/src/block_sync.rs:
```diff
@@ -490,50 +519,51 @@ impl BlockDownloader {
 				Err(BlockImportError(BlockImportErrorKind::Queue(QueueErrorKind::Full(limit)), _)) => {
-					debug!(target: "sync", "Block import queue full ({}), restarting sync", limit);
+					debug_sync!(self, "Block import queue full ({}), restarting sync", limit);
+					err = Some(CollectBlocksError::QueueFull);
```
Returning an error here in case of the queue being full will trigger the sync to restart

ascjones commented on ethcore/sync/src/chain/handler.rs:
```diff
@@ -292,7 +292,9 @@ impl SyncHandler {
-		if !sync.reset_peer_asking(peer_id, PeerAsking::BlockBodies) {
+		let allowed = sync.peers.get(&peer_id).map(|p| p.is_allowed()).unwrap_or(false);
+
+		if !sync.reset_peer_asking(peer_id, PeerAsking::BlockBodies) || !allowed {
```
Prevents expired body requests being handled, was causing retraction.

ascjones commented on ethcore/sync/src/chain/mod.rs:
```diff
@@ -762,22 +762,20 @@ impl ChainSync {
-						io.chain().block_total_difficulty(BlockId::Number(d.last_imported_block_number()))
-					});
+					// Only ask for old blocks if the peer has an equal or higher difficulty
+					let equal_or_higher_difficulty = peer_difficulty.map_or(false, |pd| pd >= syncing_difficulty);
```
Post fork block we were getting some dodgy blocks from some nodes, so tightening this back up. Still addresses https://github.com/paritytech/parity-ethereum/issues/9225

ascjones commented on ethcore/sync/src/chain/mod.rs:
```diff
@@ -855,14 +853,29 @@ impl ChainSync {
+								debug!(target: "sync", "Reset peer asking OldBlocks {:?}", pid);
+							}
+						}
+						downloader.reset();
```
This fixes the issue of stale responses causing the retraction after reset. Peer reset must happen before download reset.

andresilva commented on ethcore/sync/src/block_sync.rs:
```diff
@@ -292,22 +312,31 @@ impl BlockDownloader {
-					trace!(target: "sync", "No useful headers");
+					trace_sync!(self, "No useful headers, expected hash {:?}", expected_hash);
+					if let Some(eh) = expected_hash {
+						if self.blocks.heads_len() > 1 && !self.useless_expected_hashes.insert(eh) {
```
I'm not entirely sure I understand this. This is to avoid importing the bad #57 block from which we won't be able to progress. But when we reset the sync it means we'll clear the `useless_expected_hashes`, what's preventing us from getting into the same bad fork again?

ascjones replied:
The idea is that it should eventually find the correct blocks - it might try and reset a couple of times but once it gets a good peer it will move past the bad fork.

I clear `useless_expected_hashes` because I don't want to block further progress on a valid chain. It's possible to receive consecutive useless headers for some other reason other than being on the bad fork. The thinking is to be slightly overzealous on detecting being stuck (i.e. only 2 consecutive useless responses) but allowing it to keep trying again. Again on the assumption that it will retract and find the canon chain eventually. Happy to hear any better ideas though!

ascjones replied:
The check is to prevent unnecessary resets when we're at the head of the chain (i.e. with just one 'subchain' head) with `NewBlocks` sync, which was happening often. I believe this is harmless anyway, but my approach here was to try and reset only when necessary.

The counter would probably work and is slightly simpler, but again might force some unnecessary resets.

jimpo replied:
That makes sense. Can you leave a comment in the code here explaining that (the purpose of the `heads_len() > 1` check?). And perhaps it should check if the expected_hash is the last entry in `heads`, even if `heads_len()` is greater than 0.

If you think tracking useless responses by subchain head is preferable to a single counter, might I suggest changing `useless_expected_hashes` to a `H256FastMap<usize>` which tracks the number of useless responses per hash. That way, the reset threshold can be raised from 2 if necessary.

Regarding the counter,

andresilva commented on ethcore/sync/src/block_sync.rs:
```diff
@@ -549,4 +579,127 @@ impl BlockDownloader {
+	use tests::snapshot::TestSnapshotService;
+
+	use super::*;
+	// use chain::tests::{
```
Remove this?

andresilva commented on ethcore/sync/src/chain/handler.rs:
```diff
@@ -292,7 +292,9 @@ impl SyncHandler {
 			.and_then(|p| p.block_set)
 			.unwrap_or(BlockSet::NewBlocks);
-		if !sync.reset_peer_asking(peer_id, PeerAsking::BlockBodies) {
+		let allowed = sync.peers.get(&peer_id).map(|p| p.is_allowed()).unwrap_or(false);
```
Note to self: `is_allowed()` will check for require expiration.

andresilva commented on ethcore/sync/src/chain/mod.rs:
```diff
@@ -855,14 +853,29 @@ impl ChainSync {
-				} else if self.old_blocks.as_ref().map_or(false, |downloader| { downloader.is_complete() }) {
+				let mut is_complete = false;
+				if let Some(downloader) = self.old_blocks.as_mut() {
+					if let Err(err) = downloader.collect_blocks(io, false) {
```
We used to restart on `DownloaderImportError::Invalid`. Do we still want to do this?

ascjones replied:
`DownloaderImportError::Invalid` was the only error instance returned, so it was effectively a bool return. Related: https://github.com/paritytech/parity-ethereum/pull/9531/commits/48e548359504bed3c8d863263e8b608e5a190826

ascjones commented on ethcore/sync/src/chain/mod.rs:
```diff
@@ -855,14 +853,29 @@ impl ChainSync {
+						// reset in flight requests in order to prevent them being handled in the next round
+						for (pid, ref mut p) in &mut self.peers {
+							if p.block_set == Some(BlockSet::OldBlocks) {
+								p.reset_asking();
```
Addressed in https://github.com/paritytech/parity-ethereum/pull/9531/commits/48e548359504bed3c8d863263e8b608e5a190826. But have used the `DownloadAction` enum we already have instead of bool.

ascjones commented:
I've tested 4-5 fresh syncs successfully on mainnet, and also 4-5 times on POA network so this looks good.

ddorgan commented:
@ascjones looks like these test jobs aren't finishing .... does it work for you locally?

ascjones commented:
@ddorgan looks like my last commit has caused the tests to hang. Looking.

ddorgan commented:
@ascjones was wrong about that .... looks like the tests are running ok it's just taking a long time.... currently the four jobs at nearly three hours running, the normal test time (from say another build this morning on the same system) is 20 minutes.

jimpo commented on ethcore/sync/src/block_sync.rs:
```diff
@@ -490,53 +513,50 @@ impl BlockDownloader {
 			// complete sync round
-			trace!(target: "sync", "Sync round complete");
+			trace_sync!(self, "Sync round complete");
 			self.reset();
```
`reset()` doesn't need to be called here now since it's called in `ChainSync::collect_blocks`.

ascjones replied:
I did this but it was causing tests to hang and the sync to break. I need to do some debugging as to why that is, but at the moment it seems that you can't reset the peers after the sync round is complete.

jimpo commented on ethcore/sync/src/chain/mod.rs:
```diff
@@ -855,14 +853,29 @@ impl ChainSync {
-				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == Err(DownloaderImportError::Invalid) {
+				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == DownloadAction::Reset {
+					trace!(target: "sync", "Restarting NewBlocks download");
 					self.restart(io);
```
In the case of a successfully completed sync round, this probably shouldn't `restart()`? Why not handle it similar to OldBlocks by expiring peer requests and calling `reset()`?

ascjones replied:
Indeed we shouldn't. I've reverted the change which returns `DownloadAction::Reset` in case of successful round completion. 

I also agree this could be unified. My thinking for this PR was to minimize changes that might break `NewBlocks` sync, and just get something to fix ancient blocks.

jimpo commented on ethcore/sync/src/chain/mod.rs:
```diff
@@ -855,14 +853,29 @@ impl ChainSync {
 			},
 			BlockSet::OldBlocks => {
-				if self.old_blocks.as_mut().map_or(false, |downloader| { downloader.collect_blocks(io, false) == Err(DownloaderImportError::Invalid) }) {
-					self.restart(io);
```
Should it restart in the case of a bad old block?

ascjones replied:
Yes and that behaviour is retained.

jimpo replied:
Now it resets peer requests and the downloader state instead of calling `restart()`, so it looks like the behavior after an invalid ancient block has changed. But I think the new logic is better (it doesn't mess with the new blocks downloader state).

ascjones replied:
You're right, I misunderstood that comment last night. Yes that was intentional to stop the NewBlocks downloader being reset. I think that similarly the OldBlocks download perhaps shouldn't be reset (via restart) when we get a bad block for NewBlocks. Plenty of improvements still to make here!

ascjones commented:
@jimpo #9654 partially addresses some of the issues here, so we need to coordinate how we merge these. 

But as it is this set of changes does succeed in doing a full mainnet sync, run 7 times now. So I believe this is ready to be merged.

If the other one gets in first I will just rebase off that and remove the 'consecutive useless headers' stuff.

Either way I think we should try and get it in soon since it is currently broken.

ascjones commented:
@jimpo have implemented your suggestions with the useless counter and the reset logic. Have run it 3 times against mainnet and it manages to get past the corrupt blocks. Also ran a couple of full syncs against POA network and it works okay.

andresilva commented:
@ascjones Are the changes to the tests submodule expected? (maybe you updated the submodule unknowingly?)

ascjones commented:
Yes just noticed that too, I must've done that by accident. Fixing.
